src/platform/uefi_aarch64/main.zig

```diff
@@ -0,0 +1,369 @@
+                .RuntimeServicesCode,
+                => .rwx,
+                else => continue,
+            };
+
+            sabaton.paging.map(e.physical_start, e.physical_start, e.number_of_pages * page_size, perms, memory_type, root);
+            sabaton.paging.map(sabaton.upper_half_phys_base + e.physical_start, e.physical_start, e.number_of_pages * page_size, perms, memory_type, root);
+        }
+    }
+
+    fn init() @This() {
+        var result: @This() = undefined;
+        result.memory_map.ptr = @alignCast(8, sabaton.vital(allocator.alloc(u8, memory_map_size), "Allocating for UEFI memory map", true).ptr);
+        result.fetch();
+        return result;
+    }
+};
+
+pub fn main() noreturn {
+    if (locateProtocol(uefi.protocols.SimpleTextOutputProtocol)) |proto| {
+        conout = proto;
+    }
+
+    page_size = sabaton.paging.detect_page_size();
+
+    // Find RSDP
+    @import("acpi.zig").init();
+
+    // Find the root FS we booted from
+    const root = findFSRoot();
+
+    const kernel_file_bytes = sabaton.vital(fs.loadFile(root, "kernel.elf"), "Loading kernel ELF (esp\\kernel.elf)", true);
+
+    // Create the stivale2 tag for the kernel ELF file
+    sabaton.kernel_file_tag.kernel_addr = @ptrToInt(kernel_file_bytes.ptr);
+    sabaton.add_tag(&sabaton.kernel_file_tag.tag);
+
+    var kernel_elf_file = sabaton.Elf{
+        .data = kernel_file_bytes.ptr,
+    };
+    kernel_elf_file.init();
+
+    var kernel_stivale2_header: sabaton.Stivale2hdr = undefined;
+    _ = sabaton.vital(
+        kernel_elf_file.load_section(".stivale2hdr", sabaton.util.to_byte_slice(&kernel_stivale2_header)),
+        "loading .stivale2hdr",
+        true,
+    );
+
+    const kernel_memory_bytes = sabaton.vital(allocator.alignedAlloc(u8, 0x1000, kernel_elf_file.paged_bytes()), "Allocating kernel memory", true);
+
+    // Prepare a paging root for the kernel
+    paging_root = sabaton.paging.init_paging();
+
+    // Load the kernel into memory
+    kernel_elf_file.load(kernel_memory_bytes, &paging_root);
+
+    // Ought to be enough for any firmwares crappy memory layout, right?
+    const stivale2_memmap_bytes = @alignCast(8, sabaton.vital(allocator.alloc(u8, 64 * 1024), "Allocating for stivale2 memory map", true));
+
+    // Get a framebuffer
+    @import("framebuffer.zig").init();
+
+    // Get the memory map to calculate a max address used by UEFI
+    var memmap = MemoryMap.init();
+
+    memmap.map_everything(&paging_root);
+
+    // Now we need a memory map to exit boot services
+    memmap.fetch();
+
+    memmap.parse_to_stivale2(stivale2_memmap_bytes);
+    sabaton.add_tag(@ptrCast(*sabaton.Stivale2tag, stivale2_memmap_bytes.ptr));
+
+    uefiVital(uefi.system_table.boot_services.?.exitBootServices(uefi.handle, memmap.key), "Exiting boot services");
+
+    conout = null; // We can't call UEFI after exiting boot services
+
+    sabaton.paging.apply_paging(&paging_root);
+
+    asm volatile (
+        \\  MSR SPSel, XZR
```

Comment:
Maybe refactor this into a separate function

Reply:
Fair